docs/exceptions.md

```diff
@@ -0,0 +1,246 @@
+
+* return `Either ErrorADT`, `Maybe`
+* wrap the underlying (pure!) monad in `ExceptT` or `CatchT`
+* use `MonadError` or `MonadThrow`
+
+Consider parsing: it is pure, but we cannot make assumptions about the input. In
+this case we might want to use `ExceptT ParseError`. Or consider a lookup in a
+`Map`, where we don't know whether the key is present -- in this case we'd like
+to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
+
+* `Maybe`, `Either e`
+* `MaybeT`, `ExceptT e`
+* `CatchT`
+
+Avoid using `Text` with the error message in place of `e` -- create a proper
+ADT. In case creating a proper ADT feels too cumbersome, use `CatchT`, which
+is equivalent to `ExceptT SomeException`.
+
+Be careful not to use `MaybeT`, `ExceptT`, and `CatchT` in potentially impure
+code. When in doubt whether the code is potentially impure, use `MonadThrow`.  
+(The reason we don't want `ExceptT` and co. in potentially impure code is that
+they add additional exception mechanisms to the one that `IO` has, and
+`catch`/`bracket` don't account for this).
+
+### Impure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `ExceptT`, `MaybeT`, or `CatchT`
+* use `Maybe` or `Either`
+* use `MonadError`
+* use `throwIO`
+
+Do:
+
+* create a custom exception type
+* use `throwM` (`MonadThrow`)
+
+We disallow the use of `throwIO` only because it is redundant in the presence of
+`throwM` and requires a stronger constraint (`MonadIO` rather than
+`MonadThrow`).
+
+Derive prisms for exception types with multiple constructors, so it's convenient
+to use them with `catchJust`.
+
+### Impure code, programmer mistakes
+
+Use the same techniques as in pure code -- `error` or `impureThrow`. There are
+two reasons for this:
+
+* we would rather catch the error sooner than later, and `impureThrow` explodes
+  when forced to WHNF, while `throwIO` or `throwM` explode when executed
+
+* when in an abstract (but potentially impure) monad, using `throwM` might
+  add an additional constraint
+
+## Packages and modules
+
+Do not import `Control.Exception` or `Control.Monad.Catch`! We use the
+`safe-exceptions` to deal with asynchronous exceptions gracefully, so import
+`Control.Exception.Safe`.
+
+## Resource handling
+
+Use `bracket` or `ResourceT` to guarantee the release of resources. In case of
+concurrent code, avoid `forkIO` or `forkProcess` in favor of the `async`
+package, as it rethrows exceptions from the child threads.
+
+## Migration
+
+We should get rid of `Mockable Throw` and `Mockable Catch`, as they buy us
+nothing compared to `MonadThrow` and `MonadCatch` but have less ecosystem
+support -- for instance, the `safe-exceptions` package doesn't use them.
+
+We should identify the parts of the code that use `ExceptT` in impure or
+potentially impure code and replace them with exceptions.
+
+We should make sure that no code imports `Control.Exception` or
+`Control.Monad.Catch`, and use `Control.Exception.Safe` instead.
+
```

Comment:
I wrote few more things which I want to be done. But I admit they might take quite a lot of time and I don't know how much time we can spend on this Epic.
If I had to choose only the most important points from those, I would choose these three 👍 
But I think it'd be good to also add other improvements which I suggested in comments to CSL-1842 (avoiding `Text` for errors, avoiding exceptions leakages due to `fork`s, etc.). Maybe we don't have time to do them within this Epic, but at least it will provide broader picture of the problems we currently have.

Reply:
> avoiding Text for errors, avoiding exceptions leakages due to forks, etc

I included these in the guidelines, but not in the migration plan. I'll make it more explicit and copy these to the migration plan.